### ISSUE TITLE
the recog pred is none after training my custom data

### ISSUE THREAD

**Issue by LolaWei** [open-mmlab/mmocr#760]
**I have trained my custom data based on recog- CRNN, then I want to use ocr.py to predict some imgaes using the following codes:**
`img_path = '/disks/sde/lola/OCR/MMOCR/models/imgs'`
`recog_config = 'configs/textrecog/crnn/crnn_en_essay_no_pretrain_.py'`
`recog_ckpt = '/disks/sde/lola/OCR/MMOCR/models/crnn_en_essay/best_0_char_precision_epoch_6.pth'`
`ocr = MMOCR(det=None, recog_config=recog_config, `recog_ckpt=recog_ckpt,imshow=True)`
`results = ocr.readtext(img_path, details=True, print_result=True)`

**But the result is blank:**

load checkpoint from local path: /disks/sde/lola/OCR/MMOCR/models/crnn_en_essay/best_0_char_precision_epoch_6.pth
{'text': '', 'score': 0.0}

{'text': '', 'score': 0.0}

{'text': '', 'score': 0.0}

{'text': '', 'score': 0.0}

{'text': '', 'score': 0.0}

{'text': '', 'score': 0.0}

however, the char _acc of this model is 0.8333, i am wondering if  i use the wrong code to predict recog results

**Comment by gaotongxiao** [comment]
It's possible that your model is either underfitting or overfitting to your training set and can't generalize well on new pictures. The simplest way to check the correctness of your steps is to feed some images from the training set and look at the results.

Also, what were the full evaluation results on your best model? It looks like you saved your best model based on the character precision metric. However, it's not a good indicator of the recognizer's performance compared to `word_acc_ignore_case_symbol`, unless that's exactly what you need. Here are the explanations for these metrics:
https://github.com/open-mmlab/mmocr/blob/b8f7ead74cb0200ad5c422e82724ca6b2eb1c543/mmocr/core/evaluation/ocr_metric.py#L92-L103

Hence, ocr.py matches the text

**Comment by LolaWei** [comment]
thanks for your reply!!
1. I predicted some imgaes from train dataset. still I got nothing....
2. my dataset are handwrithing english images like the following, so i always used NED as the indicator, i am not sure in MMOCR which indicator will be the best.
![14_0202100101_76_7](https://user-images.githubusercontent.com/56593857/151314340-df798963-286b-45b6-ba73-e0d1ea6a77f1.jpg)
 3. i now swiitched the indicator you suggest, and here are some of my logs, but the indicator continues to be very low.....i dont know why, because i used to use PPOCR and EASYOCR to train the same dataset, the NED are all between 80%-90%
 
[20220127_152123.log](https://github.com/open-mmlab/mmocr/files/7948591/20220127_152123.log)

**Comment by gaotongxiao** [comment]
Your model doesn't seem to converge. There are some long pictures in your dataset containing several words separated by spaces. However, MMOCR can't recognize spaces directly, unless you use some tricks to ensure your annotation is still valid (https://github.com/open-mmlab/mmocr/issues/725).

**Comment by gaotongxiao** [comment]
Let's move the discussion back to this issue. Would you share your new log after you've made changes to space cases? At least from the log you've shared, the spaces were not well-treated. And could you post 2-3 samples of your training set and their corresponding annotations? I can conduct some local tests before getting back to you with these.

**Comment by LolaWei** [comment]
ok，  of course!
1. this is the new log. since the indicator continues to be low, so i killed the training process, the log is incompelete
[20220127_152123.log](https://github.com/open-mmlab/mmocr/files/7951738/20220127_152123.log)
2. samples of my training set:
![14_0202100101_76_4](https://user-images.githubusercontent.com/56593857/151392923-6e2db7a3-0fb9-45bb-be0e-a0d491f09bdf.jpg)
![14_0202100102_76_0](https://user-images.githubusercontent.com/56593857/151392971-d7d61861-a921-45ad-83d2-9f20ec24c9d2.jpg)
![14_0202100105_76_3](https://user-images.githubusercontent.com/56593857/151392980-d850cb92-bd21-439d-8763-9c6384547789.jpg)
[rec_train_labels.txt](https://github.com/open-mmlab/mmocr/files/7951783/rec_train_labels.txt)
3.config files:
**-  based on crnn**
_base_ = [
    '../../_base_/default_runtime.py', '../../_base_/recog_models/crnn.py',
    '../../_base_/recog_pipelines/crnn_pipeline.py',
    #'../../_base_/recog_datasets/MJ_train.py',
    #'../../_base_/recog_datasets/academic_test.py',
     '../../_base_/recog_datasets/en_essay_no_pretrain.py', # config file of train data
     '../../_base_/recog_datasets/en_essay_no_pretrain_test.py', # config file of test data
    '../../_base_/schedules/schedule_adadelta_5e.py'
]
# 添加词典

dict_file = '/disks/sde/lola/OCR/MMOCR/mmocr/ppocr_keys_v1.txt' # 指定词典的地址

label_convertor = dict(
    type='CTCConvertor',dict_file=dict_file, with_unknown=True) # 把字典转换成数字 char

model = dict(
    type='CRNNNet',
    preprocessor=None,
    backbone=dict(type='VeryDeepVgg', leaky_relu=False, input_channels=1),
    encoder=None,
    decoder=dict(type='CRNNDecoder', in_channels=512, rnn_flag=True),
    loss=dict(type='CTCLoss'),
    label_convertor=label_convertor)

train_list = {{_base_.train_list}}
test_list = {{_base_.test_list}}

train_pipeline = {{_base_.train_pipeline}}
test_pipeline = {{_base_.test_pipeline}}

# test If here to asign epoch
total_epochs = 100

data = dict(
    samples_per_gpu=64,
    workers_per_gpu=4,
    val_dataloader=dict(samples_per_gpu=10),
    test_dataloader=dict(samples_per_gpu=10),
    train=dict(
        type='UniformConcatDataset',
        datasets=train_list,
        pipeline=train_pipeline),
    val=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline),
    test=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline))

#evaluation = dict(interval=1, metric='acc',save_best = "ocr_metric:ocr_metric",rule="greater") # save best
evaluation = dict(interval=1, metric='acc',save_best="0_word_acc_ignore_case_symbol",rule="greater")
cudnn_benchmark = True

**train-dataset :**
# Text Recognition Training set, including:
# Synthetic Datasets: Syn90k

# 这里得把trian_images替换一下
train_root = '/disks/sde/lola/OCR/PPOCR/train_data/en_essay'

#train_img_prefix = f'{train_root}/mnt/ramdisk/max/90kDICT32px'
#train_ann_file = f'{train_root}/label.lmdb'

# change it to the following: dont need dir train
train_img_prefix = f'{train_root}/' # 会自动从label中读取上一级目录的
train_ann_file = f'{train_root}/rec_train_labels.txt'

train = dict(
    type='OCRDataset',
    img_prefix=train_img_prefix,
    ann_file=train_ann_file,
    loader=dict(
        #type='LmdbLoader',
         type='HardDiskLoader',
        repeat=1,
        parser=dict(
            type='LineStrParser',
            keys=['filename', 'text'],
            keys_idx=[0, 1],
            separator='\t')),
    pipeline=None,
    test_mode=False)

train_list = [train]

**test-dataset**
# Text Recognition Testing set, including:
# Regular Datasets: IIIT5K, SVT, IC13
# Irregular Datasets: IC15, SVTP, CT80

#test_root = 'data/mixture'
test_root = '/disks/sde/lola/OCR/PPOCR/train_data/en_essay'
# test image files
# test_img_prefix1 = f'{test_root}/III5K/'
test_img_prefix7 = f'{test_root}/'
# test ann files
test_ann_file1 = f'{test_root}/III5K/test_labels.txt'
test_ann_file7 = f'{test_root}/rec_test_labels.txt'

# test1 = dict(
#     type='OCRDataset',
#     img_prefix=test_img_prefix1,
#     ann_file=test_ann_file1,
#     loader=dict(
#         type='HardDiskLoader',
#         repeat=1,
#         parser=dict(
#             type='LineStrParser',
#             keys=['filename', 'text'],
#             keys_idx=[0, 1],
#             separator=' ')),
#     pipeline=None,
#     test_mode=True)

test7 = dict(
    type='OCRDataset',
    img_prefix=test_img_prefix7,
    ann_file=test_ann_file7,
    loader=dict(
        type='HardDiskLoader',
        repeat=1,
        parser=dict(
            type='LineStrParser',
            keys=['filename', 'text'],
            keys_idx=[0, 1],
            separator='\t')), # 注意这里的适配性,是tab键
    pipeline=None,
    test_mode=True)
#test_list = [test1, test2, test3, test4, test5, test6]
test_list = [test7] #use test7-en essay to test

**Comment by LolaWei** [comment]
here are the config files, thanks very much!!
[crnn_en_essay_no_pretrain_.txt](https://github.com/open-mmlab/mmocr/files/7951865/crnn_en_essay_no_pretrain_.txt)

**Comment by gaotongxiao** [comment]
Some problems I've found:
1. Your label converter loads the dictionary from a file, however a space in the line will still be ignored during parsing. Therefore, you should set up CTCConvertor like the following:
```python
label_convertor = dict(
    type='CTCConvertor',
    dict_list=list('0123456789abcdefghijklmnopqrstuvwxyz '),
    with_unknown=True,
    lower=True)
```
2. The training samples all seem to be quite long, whereas the default CRNN only recognizes no more than 9 characters at a time. To remedy the situation, you should figure out the maximum number of characters that CRNN will be recognizing, with which you calculate the desired width of the resize image. (in CRNN, the maximum number of characters depends on the width of the input image)
Concretely, if the maximum number of output characters is 70, the width of your input image should be (70-1)*4=276 (find the formula in the [docs of CRNN's backbone](https://mmocr.readthedocs.io/en/latest/api.html#mmocr.models.textrecog.backbones.VeryDeepVgg)). Then, change the parameters of `ResizeOCR` in `train_pipeline` and `test_pipeline`:
```python
    dict(
        type='ResizeOCR',
        height=32,
        min_width=276,
        max_width=276,
        keep_aspect_ratio=False),
```
Then CRNN can generate up to 70 characters.

3. Although technically feasible, training CRNN on such long sequences are not recommended. You may consider truncating your  examples for faster convergence and better performance. Since generating long sequences can be difficult, you should also consider increasing the number of training epochs and tune the model parameters.
4. Given the difficulty of fully matching long sequences, `word_acc_ignore_case_symbol` is probably not a good indicator for your task

**Comment by LolaWei** [comment]
thanks for your detailed explanation! I will try change these configs furture~  finally, 新年快乐！谢谢你的耐心解答！

**Comment by LolaWei** [comment]
@gaotongxiao  
Hi! though I still failed to get handwriting english results, but I succeed in chinese handwriting dataset following your above instructions! and the o-char_precision has reached to 92%,but I still have two questions:
1. what's the diference between char_precision and N_E_D?
2. as you have seen before, can you give me some suggestions on training such data？for example, which algorithm can I try on my dataset in order to get better results?

**Comment by gaotongxiao** [comment]
1. char_precision is the average percentage of predicted characters that correctly match the ground truth texts. 1-N.E.D is an official metric used in the ICDAR competition where N.E.D stands for Normalized Edit Distance. The motivation of adaptation is that "_Chinese script has much more and also usually longer vocabulary than Latin script, which makes word accuracy metric too harsh to proper evaluate T2.2_". (read https://rrc.cvc.uab.es/?ch=14&com=tasks). And it's better to read through their implementations at https://github.com/open-mmlab/mmocr/blob/b8f7ead74cb0200ad5c422e82724ca6b2eb1c543/mmocr/core/evaluation/ocr_metric.py#L84 to get a full understanding.
2. Unfortunately, we had no experience with training recognizers on long pictures. You might try out some latest models, such as SATRN or ABINet, but don't forget to change the maximum number of output characters accordingly. (e.g., the parameter in SATRN is at https://github.com/open-mmlab/mmocr/blob/0a1787d6bcbec25649441a1f26e928ea7158e4c4/configs/textrecog/satrn/satrn_academic.py#L41 )

**Comment by gaotongxiao** [comment]
I'm closing this issue, but feel free to reopen it if you have any followup questions.
